File: src/runtime/boxes/brk/bb_brk.java
package bb;

/**
 * bb_brk.java — BRK (BREAK): scan to first char in set (zero-width possible)
 * Port of bb_brk.c / bb_brk.s
 *
 *   BRK_α:  δ=0; while (Δ+δ<Ω && !chars.has(Σ[Δ+δ])) δ++;
 *           if (Δ+δ >= Ω)               goto BRK_ω;   (never found char in set)
 *           BRK=spec(Σ+Δ,δ); Δ+=δ;     goto BRK_γ;
 *   BRK_β:  Δ-=δ;                       goto BRK_ω;
 */
public class bb_brk extends bb_box {
    private final String chars;
    private int          delta;

    public bb_brk(MatchState ms, String chars) { super(ms); this.chars=chars; }

    @Override public Spec α() {
        delta = 0;
        while (ms.delta + delta < ms.omega &&
               chars.indexOf(ms.sigma.charAt(ms.delta + delta)) < 0) delta++;
        if (ms.delta + delta >= ms.omega) return null;
        Spec r = new Spec(ms.delta, delta);
        ms.delta += delta;
        return r;
    }

    @Override public Spec β() { ms.delta -= delta; return null; }
}

File: src/runtime/boxes/brk/bb_brk.cs
// bb_brk.cs — BREAK: scan to first char in set (zero-advance is OK)
// Mirrors src/runtime/boxes/brk/bb_brk.c
//
// α: scan forward while char NOT in charset;
//    if reached EOS without finding a set-char → ω;
//    cursor += count; γ(count chars)
// β: cursor -= count; ω

namespace Snobol4.Runtime.Boxes;

public sealed class bb_brk : IByrdBox
{
    private readonly string _chars;
    private int _count;

    public bb_brk(string chars) { _chars = chars ?? ""; }

    public Spec Alpha(MatchState ms)
    {
        _count = 0;
        while (ms.Cursor + _count < ms.Length &&
               !ms.CharInSet(ms.Cursor + _count, _chars))
            _count++;
        // must have found a set-char (not hit EOS)
        if (ms.Cursor + _count >= ms.Length) return Spec.Fail;
        var result = Spec.Of(ms.Cursor, _count);
        ms.Cursor += _count;
        return result;
    }

    public Spec Beta(MatchState ms)
    {
        ms.Cursor -= _count;
        return Spec.Fail;
    }
}
